Add a documentation object to main eslint

ABaldwinHunter opened the pull request:
This change aims to make markdown [documentation](https://github.com/eslint/eslint/tree/master/docs/rules) for each issue more accessible for generating content. 

Implementation:
1) update `package.json` to include docs as part of package
2) add a `docs` function that returns an object containing rule name and markdown content.

<img width="1124" alt="screen shot 2015-10-17 at 1 11 46 am" src="https://cloud.githubusercontent.com/assets/8718443/10557241/bd556ca4-7471-11e5-8d83-2cbc537b689b.png">

you can access markdown content by checkname:

<img width="1132" alt="screen shot 2015-10-17 at 1 53 46 am" src="https://cloud.githubusercontent.com/assets/8718443/10557249/0b852bd0-7472-11e5-828a-05ee8af8b078.png">

cc @codeclimate/review @GordonDiggs

ABaldwinHunter commented:
cc @mrb @jpignata

jpignata commented:
Is your thought here that you'd open a pull into eslint or that we'd maintain a fork of eslint?

wfleming commented on lib/docs.js:
```diff
@@ -0,0 +1,19 @@
+function Docs() {
+  var fs = require('fs'),
+      path = require('path'),
+      rules = fs.readdirSync(path.resolve('node_modules/eslint/docs/rules')),
```
Can we determine & resolve this path relative to [`__dirname`](https://nodejs.org/docs/latest/api/globals.html#globals_dirname)? I think this would be less fragile (and easier to test).

Also, I know I'm a broken record, but can we write tests? :neckbeard:

ABaldwinHunter replied:
@wfleming Tests sound good!

Sent from my iPhone

> On Oct 17, 2015, at 11:16 AM, Will Fleming notifications@github.com wrote:
> 
> In lib/docs.js:
> 
> > @@ -0,0 +1,19 @@
> > +// Create object for providing easy access to documentation
> > +
> > +function Docs() {
> > -  var fs = require('fs'),
> > -      path = require('path'),
> > -      rules = fs.readdirSync(path.resolve('node_modules/eslint/docs/rules')),
> >   Can we determine & resolve this path relative to __dirname? I think this would be less fragile (and easier to test).
> 
> Also, I know I'm a broken record, but can we write tests? 
> 
> —
> Reply to this email directly or view it on GitHub.

wfleming commented:
Great feature to add :+1:, @ABaldwinHunter. Nice of them to already keep these things in the repository so we don't need to jump through the same hoops as with brakeman.

ABaldwinHunter commented:
My thought might be to use our fork in short term and open a pull against ESLint in long term, since this change seems potentially useful to others but might take a while to get merged.

My understanding is that we currently use a fork of ESLint to avoid plugins. 

I talked to @GordonDiggs a bit, and not straying too far from the ESLint tool feels important.

For an ESLint PR, I would do a little more investigation to understand the most desirable way to make a change. I like will's idea of adding tests, and also wonder if it'd be better to not pollute the space of objects returned by main (require('eslint')). Although having the 'docs' function is convenient for our purposes, having markdown docs available by adding 'docs' to 'files' listed in package.json would be sufficient for us to keep the reading function internal to codeclimate-ESLint . 

Sent from my iPhone

> On Oct 17, 2015, at 9:45 AM, John Pignata notifications@github.com wrote:
> 
> Is your thought here that you'd open a pull into eslint or that we'd maintain a fork of eslint?
> 
> —
> Reply to this email directly or view it on GitHub.

ABaldwinHunter commented:
Wdyt? Interested in other thoughts 

Sent from my iPhone

> On Oct 17, 2015, at 9:45 AM, John Pignata notifications@github.com wrote:
> 
> Is your thought here that you'd open a pull into eslint or that we'd maintain a fork of eslint?
> 
> —
> Reply to this email directly or view it on GitHub.

jpignata commented:
Got it. I missed the detail that we're already maintaining a fork.

jpignata commented:
:+1:! This feels like the right approach to me.

gdiggs committed:
Add a documentation module.

This provides access to the Markdown source for documentation of each
rule. It is used as follows:

    var docs = require("docs");
    docs.get("complexity");

Special mention to @ABaldwinHunter who did the basis of this work.

gdiggs commented:
@codeclimate/review ready for another look! I modeled this after the [rules module](https://github.com/codeclimate/eslint/blob/master/lib/rules.js) to match ESLint's style.

BlakeWilliams commented on lib/docs.js:
```diff
@@ -0,0 +1,38 @@
+var fs = require("fs"),
+    path = require("path");
+
+//------------------------------------------------------------------------------
```
Can we remove the banner comments? Maybe replace them with documentation?

gdiggs replied:
These are standard within ESLint - I added them to match the repo's style

BlakeWilliams replied:
Gotcha, that's unfortunate.

BlakeWilliams commented:
LGTM
